sbibm/tasks/norefposterior/task.py

```diff
@@ -0,0 +1,365 @@
+from pathlib import Path
+from typing import Any, Callable, Dict, Optional
+
+import pyro
+import torch
+from pyro import distributions as pdist
+
+from sbibm import get_logger
+from sbibm.tasks.simulator import Simulator
+from sbibm.tasks.task import Task
+from sbibm.utils.io import get_tensor_from_csv, save_tensor_to_csv
+
+
+def torch_average(a, weights=None, axis=0):
+
+    if isinstance(weights, type(None)):
+        return a.mean(axis=axis)
+    else:
+        assert weights.sum() > 0, f"received all 0 weights tensor"
+        value = torch.sum(a * weights, axis=axis) / torch.sum(weights, axis=axis)
+        return value
+
+
+def base_coordinate_field(min_axis=-16, max_axis=16):
+    """returns a torch tensor that contains the coordinates of a regular
+    grid between <min_axis> and <max_axis> broadcasted/cloned
+    <batchsize> times, i.e.
+    >>> arr = quadratic_coordinate_field(-3,3)
+    >>> arr.shape
+    (6,6,2)
+     ^^^---- dimensions of max_axis - min_axis, 3-(-3)
+    """
+    size_axis = max_axis - min_axis
+
+    x = torch.arange(min_axis, max_axis).detach().float()
+    y = torch.arange(min_axis, max_axis).detach().float()
+
+    xx, yy = torch.meshgrid(x, y)
+    val = torch.swapaxes(torch.stack((xx.flatten(), yy.flatten())), 1, 0).float()
+
+    value = val.reshape(size_axis, size_axis, 2)
+
+    return value
+
+
+def bcast_coordinate_field(base_field, num_samples):
+
+    # boadcast to <num_samples> doublicates
+    valr_ = torch.broadcast_to(base_field, (num_samples, *base_field.shape)).detach()
+
+    # move axis from position 2 to front
+    value = torch.swapaxes(valr_, 2, 0)
+
+    return value
+
+
+def quadratic_coordinate_field(min_axis=-16, max_axis=16, batch_size=32):
+    """returns a torch tensor that contains the coordinates of a regular
+    grid between <min_axis> and <max_axis> broadcasted/cloned
+    <batchsize> times, i.e.
+    >>> arr = quadratic_coordinate_field(-3,3,4)
+    >>> arr.shape
+    #    --- batch_size
+    #    v
+    (6,6,4,2)
+    #^ ^
+    #| |
+    #------- dimensions of max_axis - min_axis, 3-(-3)
+    """
+
+    valr = base_coordinate_field(min_axis, max_axis)
+
+    # at every point of the image w=size_axis x w=size_axis
```

Comment:
Into docstring of `base_coordinate_field`?

Reply:
What do you mean?

Reply:
I meant that the comment starting "at every point ..." might belong into the function `base_coordinate_field` as a docstring?

Reply:
done with ab6aa26